### PR TITLE
Add setting for enabling/disabling circuit breaker

### PR DESCRIPTION
**Pull request by kaituo**
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/143
https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/159

*Description of changes:*
A circuit breaker is broken when heap memory usage exceeds 85%, and the related AD job would be disabled because of that. It is possible at one point the heap memory usage exceeds 85% and gets back to less than 85% soon afterward.

This PR mitigates the issue in the two following ways:
First, only disable the AD job after the circuit breaker is broken for a consecutive number of times (3 times).
Second, add a setting for enabling/disabling circuit breaker.

Testing done:
1. After disabling circuit breaker, an open circuit breaker does not affect AD job execution.
2. Verified an open circuit breaker wouldn't cause an AD job to be stopped immediately.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java [hunk 1]
```diff
@@ -45,12 +45,19 @@
...
+
     private final Map<String, Setting<?>> settings = unmodifiableMap(new HashMap<String, Setting<?>>() {
         {
             /**
              * AD plugin enable/disable setting
              */
             put(AD_PLUGIN_ENABLED, Setting.boolSetting(AD_PLUGIN_ENABLED, true, NodeScope, Dynamic));
+
+            /**
+             * LTR breaker enable/disable setting
```
Comment (wnbts) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java:58]: issue. LTR seems to refer to AD.

**Comment by ylwu-amzn** [comment]
How about we set a dynamic setting for the memory threshold? If user just use one cluster mainly for anomaly detection, they should be able to increase the memory for AD.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java:58]: fixed

**Comment by kaituo** [comment]
> one general comment. Three data points before disabling seems too quick. The system should be resilient and be able to recover on itself when conditions improve, or human operators would be burdened with every system resource spike.

How about 6?

**Comment by kaituo** [comment]
> How about we set a dynamic setting for the memory threshold? If user just use one cluster mainly for anomaly detection, they should be able to increase the memory for AD.

This is beyond the scope of this PR.  Created an issue: https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/170

**Comment by wnbts** [comment]
> > one general comment. Three data points before disabling seems too quick. The system should be resilient and be able to recover on itself when conditions improve, or human operators would be burdened with every system resource spike.
> 
> How about 6?

my opinion is that this number should be reasonably large for recovering expected system spikes and on the other side, not too large that the task is needlessly retrying in a permanently/consistently inoperable system. 3 data points/3 minutes seems too short for the former. I am not sure about 6 but it seems more reasonable.

**Comment by kaituo** [comment]
> > > one general comment. Three data points before disabling seems too quick. The system should be resilient and be able to recover on itself when conditions improve, or human operators would be burdened with every system resource spike.
> > 
> > 
> > How about 6?
> 
> my opinion is that this number should be reasonably large for recovering expected system spikes and on the other side, not too large that the task is needlessly retrying in a permanently/consistently inoperable system. 3 data points/3 minutes seems too short for the former. I am not sure about 6 but it seems more reasonable.

suppose detector interval is 1 hour, then 6 means 6 hour retry before terminating models.  This value should be different among detectors, considering detector interval.  What do you think?

**Comment by wnbts** [comment]
> hen 6 means 6 hour retry before terminating models. This value should be different among detectors, considering detector interval. What do you think?

that would be ideal. if that's not easy to do, at worst 6 hours means 6 attempts, not too big a problem.

**Comment by kaituo** [comment]
> > hen 6 means 6 hour retry before terminating models. This value should be different among detectors, considering detector interval. What do you think?
> 
> that would be ideal. if that's not easy to do, at worst 6 hours means 6 attempts, not too big a problem.

I'll set it to 6 for now and created an issue https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/171 to track it.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/LimitExceededException.java [hunk 2]
```diff
@@ -29,4 +29,15 @@
     public LimitExceededException(String anomalyDetectorId, String message) {
         super(anomalyDetectorId, message, true);
     }
+
+    /**
+     * Constructor with an anomaly detector ID and an explanation.
```
Comment (wnbts) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/LimitExceededException.java:34]: minor. and flag for stopping.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java [hunk 3]
```diff
@@ -100,6 +107,10 @@ public static boolean isADPluginEnabled() {
         return EnabledSetting.getInstance().getSettingValue(EnabledSetting.AD_PLUGIN_ENABLED);
     }
 
+    public static boolean isADBreakerEnabled() {
```
Comment (wnbts) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java:110]: Minor. Java doc for public interface is missing.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/LimitExceededException.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/LimitExceededException.java:34]: added

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/AnomalyDetectorSettings.java [hunk 4]
```diff
@@ -133,7 +133,7 @@ private AnomalyDetectorSettings() {}
     public static final Setting<Integer> MAX_RETRY_FOR_END_RUN_EXCEPTION = Setting
         .intSetting(
             "opendistro.anomaly_detection.max_retry_for_end_run_exception",
-            3,
+            6,
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/AnomalyDetectorSettings.java:136]: added to the tracking issue.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/EnabledSetting.java:110]: added doc

**Pull request merged**